gui/src/graphics/vulkan.rs

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+use ash::vk::{ApplicationInfo, InstanceCreateInfo};
+use std::ffi::CStr;
+use thiserror::Error;
+
+pub struct Vulkan {
+    entry: ash::Entry,
+    instance: ash::Instance,
+    devices: Vec<VulkanPhysicalDevice>,
+}
+
+impl super::GraphicsApi for Vulkan {
+    type PhysicalDevice = VulkanPhysicalDevice;
+
+    type CreateError = VulkanCreateError;
+
+    fn new() -> Result<Self, Self::CreateError> {
+        let entry = ash::Entry::linked();
+
+        let app_info = ApplicationInfo::default().application_name(c"Obliteration");
+
+        let create_info = InstanceCreateInfo::default().application_info(&app_info);
+
+        let instance = unsafe { entry.create_instance(&create_info, None) }
+            .map_err(VulkanCreateError::CreateInstanceFailed)?;
+
+        let devices = unsafe { instance.enumerate_physical_devices() }
+            .map_err(VulkanCreateError::EnumeratePhysicalDevicesFailed)?
+            .into_iter()
+            .map(
+                |device| -> Result<VulkanPhysicalDevice, VulkanCreateError> {
+                    let properties = unsafe { instance.get_physical_device_properties(device) };
+
+                    let name = CStr::from_bytes_until_nul(unsafe {
```

Comment:
Why don't use `CStr::from_ptr` instead of making an intermediate slice?

Reply:
In theory, there could be no null bytes in the 256 bytes. This way, we can catch that.

Reply:
In that case better to remove a hard-coded length otherwise it will be UB if the length does not match in the future.